File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/query/InheritanceQueryTest.java
package org.babyfish.jimmer.sql.query;

import org.babyfish.jimmer.sql.JSqlClient;
import org.babyfish.jimmer.sql.common.AbstractQueryTest;
import org.babyfish.jimmer.sql.model.inheritance.PermissionFetcher;
import org.babyfish.jimmer.sql.model.inheritance.Role;
import org.babyfish.jimmer.sql.model.inheritance.RoleFetcher;
import org.babyfish.jimmer.sql.model.inheritance.RoleTable;
import org.babyfish.jimmer.sql.runtime.LogicalDeletedBehavior;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;

public class InheritanceQueryTest extends AbstractQueryTest {

    @Override
    protected LambdaClient getLambdaClient() {
        return getLambdaClient(it -> {
            it.setLogicalDeletedBehavior(LogicalDeletedBehavior.IGNORED);
        });
    }

    @Test
    public void testFetchLonely() {
        executeAndExpect(
                getLambdaClient().createQuery(RoleTable.class, (q, role) -> {
                    return q.select(
                            role.fetch(
                                    RoleFetcher.$
                                            .name()
                            )
                    );
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME from ROLE tb_1_"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{\"name\":\"r_1\",\"id\":100}," +
                                    "--->{\"name\":\"r_2\",\"id\":200}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void testFetchIdOnlyChildren() {
        executeAndExpect(
                getLambdaClient().createQuery(RoleTable.class, (q, role) -> {
                    return q.select(
                            role.fetch(
                                    RoleFetcher.$
                                            .name()
                                            .permissions()
                            )
                    );
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME from ROLE tb_1_"
                    );
                    ctx.statement(1).sql(
                            "select tb_1_.ROLE_ID, tb_1_.ID " +
                                    "from PERMISSION tb_1_ " +
                                    "where tb_1_.ROLE_ID in (?, ?)"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"name\":\"r_1\"," +
                                    "--->--->\"permissions\":[" +
                                    "--->--->--->{\"id\":1000},{" +
                                    "--->--->--->\"id\":2000}" +
                                    "--->--->]," +
                                    "--->--->\"id\":100" +
                                    "--->},{" +
                                    "--->--->\"name\":\"r_2\"," +
                                    "--->--->\"permissions\":[" +
                                    "--->--->--->{\"id\":3000}," +
                                    "--->--->--->{\"id\":4000}" +
                                    "--->--->]," +
                                    "--->--->\"id\":200" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void testFetchChildren() {
        executeAndExpect(
                getLambdaClient().createQuery(RoleTable.class, (q, role) -> {
                    return q.select(
                            role.fetch(
                                    RoleFetcher.$
                                            .name()
                                            .permissions(
                                                    PermissionFetcher.$.name()
                                            )
                            )
                    );
                }),
                ctx -> {
                    ctx.sql("select tb_1_.ID, tb_1_.NAME from ROLE tb_1_");
                    ctx.statement(1).sql(
                            "select tb_1_.ROLE_ID, tb_1_.ID, tb_1_.NAME " +
                                    "from PERMISSION tb_1_ " +
                                    "where tb_1_.ROLE_ID in (?, ?)"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"name\":\"r_1\"," +
                                    "--->--->\"permissions\":[" +
                                    "--->--->--->{\"name\":\"p_1\",\"id\":1000}," +
                                    "--->--->--->{\"name\":\"p_2\",\"id\":2000}" +
                                    "--->--->]," +
                                    "--->--->\"id\":100" +
                                    "--->},{" +
                                    "--->--->\"name\":\"r_2\"," +
                                    "--->--->\"permissions\":[" +
                                    "--->--->--->{\"name\":\"p_3\",\"id\":3000}," +
                                    "--->--->--->{\"name\":\"p_4\",\"id\":4000}" +
                                    "--->--->]," +
                                    "--->--->\"id\":200" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }

    @Test
    public void testQueryByTime() {
        executeAndExpect(
                getLambdaClient().createQuery(RoleTable.class, (q, role) -> {
                    q.where(
                            role.createdTime().between(
                                    LocalDateTime.of(2022, 10, 3, 0, 0, 0),
                                    LocalDateTime.of(2022, 10, 4, 0, 0, 0)
                            )
                    );
                    return q.select(role.fetch(RoleFetcher.$.allScalarFields().deleted(false)));
                }),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.CREATED_TIME, tb_1_.MODIFIED_TIME " +
                                    "from ROLE tb_1_ " +
                                    "where tb_1_.CREATED_TIME " +
                                    "between ? and ?"
                    );
                    ctx.rows(
                            "[" +
                                    "--->{" +
                                    "--->--->\"name\":\"r_1\"," +
                                    "--->--->\"createdTime\":\"2022-10-03 00:00:00\"," +
                                    "--->--->\"modifiedTime\":\"2022-10-03 00:10:00\"," +
                                    "--->--->\"id\":100" +
                                    "--->},{" +
                                    "--->--->\"name\":\"r_2\"," +
                                    "--->--->\"createdTime\":\"2022-10-03 00:00:00\"," +
                                    "--->--->\"modifiedTime\":\"2022-10-03 00:10:00\"," +
                                    "--->--->\"id\":200" +
                                    "--->}" +
                                    "]"
                    );
                }
        );
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/query/InheritanceQueryTest.kt
package org.babyfish.jimmer.sql.kt.query

import org.babyfish.jimmer.sql.kt.KSqlClient
import org.babyfish.jimmer.sql.kt.ast.expression.between
import org.babyfish.jimmer.sql.kt.common.AbstractQueryTest
import org.babyfish.jimmer.sql.kt.model.inheritance.Role
import org.babyfish.jimmer.sql.kt.model.inheritance.createdTime
import org.babyfish.jimmer.sql.kt.model.inheritance.fetchBy
import org.babyfish.jimmer.sql.runtime.LogicalDeletedBehavior
import java.time.LocalDateTime
import kotlin.test.Test

class InheritanceQueryTest : AbstractQueryTest() {

    private val _sqlClient: KSqlClient = sqlClient {
        setLogicalDeletedBehavior(LogicalDeletedBehavior.IGNORED)
    }

    @Test
    fun testFetchLonely() {
        executeAndExpect(
            _sqlClient.createQuery(Role::class) {
                select(
                    table.fetchBy {
                        name()
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME from ROLE tb_1_"""
            )
            rows(
                """[{"name":"r_1","id":100},{"name":"r_2","id":200}]"""
            )
        }
    }

    @Test
    fun testFetchIdOnlyChildren() {
        executeAndExpect(
            _sqlClient.createQuery(Role::class) {
                select(
                    table.fetchBy {
                        name()
                        permissions()
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME from ROLE tb_1_"""
            )
            statement(1).sql(
                """select tb_1_.ROLE_ID, tb_1_.ID 
                    |from PERMISSION tb_1_ 
                    |where tb_1_.ROLE_ID in (?, ?)""".trimMargin()
            )
            rows(
                """[
                    |--->{"name":"r_1","permissions":[{"id":1000},{"id":2000}],"id":100},
                    |--->{"name":"r_2","permissions":[{"id":3000},{"id":4000}],"id":200}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testFetchChildren() {
        executeAndExpect(
            _sqlClient.createQuery(Role::class) {
                select(
                    table.fetchBy {
                        name()
                        permissions {
                            name()
                        }
                    }
                )
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME from ROLE tb_1_"""
            )
            statement(1).sql(
                """select tb_1_.ROLE_ID, tb_1_.ID, tb_1_.NAME 
                    |from PERMISSION tb_1_ 
                    |where tb_1_.ROLE_ID in (?, ?)""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"name":"r_1",
                    |--->--->"permissions":[
                    |--->--->--->{"name":"p_1","id":1000},
                    |--->--->--->{"name":"p_2","id":2000}
                    |--->--->],
                    |--->--->"id":100
                    |--->},{
                    |--->--->"name":"r_2",
                    |--->--->"permissions":[
                    |--->--->--->{"name":"p_3","id":3000},
                    |--->--->--->{"name":"p_4","id":4000}
                    |--->--->],
                    |--->--->"id":200
                    |--->}
                    |]""".trimMargin()
            )
        }
    }

    @Test
    fun testQueryByTime() {
        executeAndExpect(
            _sqlClient.createQuery(Role::class) {
                where(
                    table.createdTime.between(
                        LocalDateTime.of(2022, 10, 3, 0, 0, 0),
                        LocalDateTime.of(2022, 10, 4, 0, 0, 0)
                    )
                )
                select(table)
            }
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.DELETED, tb_1_.CREATED_TIME, tb_1_.MODIFIED_TIME 
                    |from ROLE tb_1_ 
                    |where tb_1_.CREATED_TIME between ? and ?""".trimMargin()
            )
            rows(
                """[
                    |--->{
                    |--->--->"name":"r_1",
                    |--->--->"deleted":false,
                    |--->--->"createdTime":"2022-10-03 00:00:00",
                    |--->--->"modifiedTime":"2022-10-03 00:10:00",
                    |--->--->"id":100
                    |--->},{
                    |--->--->"name":"r_2",
                    |--->--->"deleted":true,
                    |--->--->"createdTime":"2022-10-03 00:00:00",
                    |--->--->"modifiedTime":"2022-10-03 00:10:00",
                    |--->--->"id":200
                    |--->}
                    |]""".trimMargin()
            )
        }
    }
}